### PR TITLE
Cast params prior to rule execution

### PR DESCRIPTION
**Pull request by manasdk**
- Action params should be cast to their proper types as described in the metadata prior to scheduling an action.
- This way actions executed from rules can be of various types.

### CODE REVIEW

File: st2common/st2common/models/utils/action_param_utils.py [hunk 1]
```diff
@@ -70,3 +79,51 @@ def is_immutable(param_meta):
     immutable_params = {k: merged_params[k] for k in immutable}
 
     return (required_params, optional_params, immutable_params)
+
+
+def cast_params(action_ref, params):
```
Comment (Kami) [inline st2common/st2common/models/utils/action_param_utils.py]: Yep, this is a much better place for this function.

File: st2common/st2common/models/utils/action_param_utils.py [hunk 2]
```diff
@@ -70,3 +79,51 @@ def is_immutable(param_meta):
...
+        parameter_schema = parameters_schema.get(k, None)
+        if not parameter_schema:
+            continue
+        parameter_type = parameter_schema.get('type', None)
+        if not parameter_type:
+            continue
+        cast = casts.get(parameter_type, None)
+        LOG.debug('Casting param: %s of type %s to type: %s', v, type(v), parameter_type)
+        if not cast:
+            continue
```
Comment (Kami) [inline st2common/st2common/models/utils/action_param_utils.py]: Hm, looks like we don't define a custom function for unicode parameters (under Python < 3) which means we will skip that parameter. We should probably handle that :)

On a related note, we should also log a message if we skip a parameter.

File: st2reactor/st2reactor/sensor/config.py [hunk 3]
```diff
@@ -21,41 +21,43 @@
 CONF = cfg.CONF
 
 
-def _register_common_opts():
-    st2cfg.register_opts()
+def _register_common_opts(ignore_errors=False):
```
Comment (Kami) [inline st2reactor/st2reactor/sensor/config.py]: Hm, do we want to pass ignore_errors to the st2cfg.register_opts?

File: st2common/st2common/models/utils/action_param_utils.py (hunk 2 above)
Comment (manasdk) [inline st2common/st2common/models/utils/action_param_utils.py]: The types we support are per json schema. I don't believe unicode is present in json-schema in the types.

However, if you are talking about how casting unicode to str if the string value is in unicode then yeah that is missed.

File: st2reactor/st2reactor/sensor/config.py (hunk 3 above)
Comment (manasdk) [inline st2reactor/st2reactor/sensor/config.py]: good catch.

**Comment by manasdk** [comment]
kami : I moved some more code around to reduce duplication and drift. see 5bcdd2728f48b6231d0698077424a339fa7f7b6d

**Comment by manasdk** [comment]
Merging in the change.

**Comment by Kami** [comment]
:+1:

**Pull request merged**